### ISSUE TITLE
NRF51822 Could Not Claim Interface (Raspberry Pi)

### ISSUE THREAD

**Issue by jcady92** [mbedmicro/pyOCD#59]
I have a Raspberry Pi running Raspbian (Debian).

I install libusb-1.0-0-dev, clone and install the pyUSB repo, and then clone and install the pyOCD.  I plug in my NRF51822-mkit and it enumerates under lsusb.  If I go into the test directory and run "sudo python basic_test.py" this is the output:

```
is_kernel_driver_active
Traceback (most recent call last):
  File "basic_test.py", line 48, in <module>
    board = MbedBoard.chooseBoard()
  File "/home/pi/pyOCD/pyOCD/board/mbed_board.py", line 161, in chooseBoard
    all_mbeds = MbedBoard.getAllConnectedBoards(transport, False, blocking, target_override)
  File "/home/pi/pyOCD/pyOCD/board/mbed_board.py", line 125, in getAllConnectedBoards
    mbed.write([0x80])
  File "/home/pi/pyOCD/pyOCD/interface/pyusb_backend.py", line 135, in write
    self.dev.ctrl_transfer(bmRequestType,bmRequest,wValue,wIndex,data)
  File "/usr/local/lib/python2.7/dist-packages/usb/core.py", line 962, in ctrl_transfer
    self._ctx.managed_claim_interface(self, interface_number)
  File "/usr/local/lib/python2.7/dist-packages/usb/core.py", line 146, in managed_claim_interface
    self.backend.claim_interface(self.handle, i)
  File "/usr/local/lib/python2.7/dist-packages/usb/backend/libusb0.py", line 468, in claim_interface
    _check(_lib.usb_claim_interface(dev_handle, intf))
  File "/usr/local/lib/python2.7/dist-packages/usb/backend/libusb0.py", line 380, in _check
    raise USBError(errmsg, ret)
usb.core.USBError: [Errno None] could not claim interface 3: Device or resource busy
```

This happens every time, though it doesn't seem that the board is claimed by anything else.  The odd thing is that interface 3 isn't the board, it's something else.  Here's the output to lsusb:

```
Bus 001 Device 002: ID 0424:9514 Standard Microsystems Corp. 
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 001 Device 003: ID 0424:ec00 Standard Microsystems Corp. 
Bus 001 Device 021: ID 0d28:0204 NXP LPC1768
```

I do have a way to get it to work properly.  If I unplug the device, then run the basic_test.py program, it outputs "INFO:root:Waiting for a USB device connected".  When I connect the device, the program runs as expected.  If I try to run the program subsequently after this, it outputs errors such as these:

```
is_kernel_driver_active
INFO:root:new board id detected: 107002161FE6E019E20F0F91
id => usbinfo | boardname
0 =>   (0xd28, 0x204) [nrf51822]
INFO:root:DAP SWD MODE initialised
[Errno None] Connection timed out
Traceback (most recent call last):
  File "basic_test.py", line 48, in <module>
    board = MbedBoard.chooseBoard()
  File "/home/pi/pyOCD/pyOCD/board/mbed_board.py", line 190, in chooseBoard
    raise e
usb.core.USBError: [Errno None] Connection timed out
```

```
is_kernel_driver_active
INFO:root:Unsupported target found: 
INFO:root:Waiting for a USB device connected
INFO:root:new board id detected: 107002161FE6E019E20F0F91
id => usbinfo | boardname
0 =>   (0xd28, 0x204) [nrf51822]
DAP_CONNECT response error
Traceback (most recent call last):
  File "basic_test.py", line 48, in <module>
    board = MbedBoard.chooseBoard()
  File "/home/pi/pyOCD/pyOCD/board/mbed_board.py", line 190, in chooseBoard
    raise e
ValueError: DAP_CONNECT response error
```

```
is_kernel_driver_active
Traceback (most recent call last):
  File "basic_test.py", line 48, in <module>
    board = MbedBoard.chooseBoard()
  File "/home/pi/pyOCD/pyOCD/board/mbed_board.py", line 161, in chooseBoard
    all_mbeds = MbedBoard.getAllConnectedBoards(transport, False, blocking, target_override)
  File "/home/pi/pyOCD/pyOCD/board/mbed_board.py", line 126, in getAllConnectedBoards
    u_id_ = mbed.read()
  File "/home/pi/pyOCD/pyOCD/interface/pyusb_backend.py", line 151, in read
    data = self.ep_in.read(self.ep_in.wMaxPacketSize, timeout)
  File "/usr/local/lib/python2.7/dist-packages/usb/core.py", line 364, in read
    return self.device.read(self, size_or_buffer, timeout)
  File "/usr/local/lib/python2.7/dist-packages/usb/core.py", line 918, in read
    self.__get_timeout(timeout))
  File "/usr/local/lib/python2.7/dist-packages/usb/backend/libusb0.py", line 507, in intr_read
    timeout)
  File "/usr/local/lib/python2.7/dist-packages/usb/backend/libusb0.py", line 562, in __read
    timeout
  File "/usr/local/lib/python2.7/dist-packages/usb/backend/libusb0.py", line 380, in _check
    raise USBError(errmsg, ret)
usb.core.USBError: [Errno None] Connection timed out
```

Unplugging and replugging it results in the could not claim device error.

Does anyone know why this is happening?

**Comment by jcady92** [comment]
Decided to try this again with all the updates 4 months later...the problem still persists.  I've duplicated it on another board (Hummingboard).  Does anyone have a solution?

**Comment by c1728p9** [comment]
Hi @jcady92 
Does running it on a desktop computer have the same problem?  I'm wondering if this could be related to #112.

**Comment by c1728p9** [comment]
If this is the same issue as #112 then it should be fixed by pull request #110.  Can anyone test this and see if the problem is fixed? It would be good to close out this issue if it has been taken care of.

**Comment by c1728p9** [comment]
Closing this issue since it is likely fixed. Please reopen if you see it again.
